shadowsocks/tcp.go

```diff
@@ -166,9 +161,29 @@ func proxyConnection(clientConn onet.DuplexConn, proxyMetrics *metrics.ProxyMetr
 	tgtConn := metrics.MeasureConn(tgtTCPConn, &proxyMetrics.ProxyTarget, &proxyMetrics.TargetProxy)
 
 	logger.Debugf("proxy %s <-> %s", clientConn.RemoteAddr().String(), tgtConn.RemoteAddr().String())
-	_, _, err = onet.Relay(clientConn, tgtConn)
-	if err != nil {
-		return onet.NewConnectionError("ERR_RELAY", "Failed to relay traffic", err)
+	fromClientErrCh := make(chan error)
+	go func() {
+		_, fromClientErr := clientConn.(io.WriterTo).WriteTo(tgtConn)
+		// Send FIN to target.
+		tgtConn.CloseWrite()
+		if fromClientErr != nil {
+			// Drain to prevent a close on cipher error.
```

Comment:
This doesn't call `clientConn.Close()`, but it does call `tgtConn.CloseWrite()`, which is likely to cause the target to close all sockets ~immediately, leading to `clientConn.ReadFrom()` returning, resulting in a call to `clientConn.CloseWrite()`.  To avoid this, I think you would have to avoid calling `clientConn.CloseWrite()` until after the drain, if `fromClientErr` is non-nil.